Scent map building optimization

Coolthulhu opened the pull request:
Scent calculations in `game.cpp` involve using public map functions that all do bounds checks and re-get the submap every single time. There are 4 bounds checks per every single tile (`has_flag`, `veh_at`, `has_flag_ter_or_furn`, `field_at`) and every tile that contains a vehicle goes through all parts of this vehicle to find obstacles. Twice.
So a 10x10 DeathRoller that has a total of 200 parts will cause a total of `2*200*100` checks per turn.

In this PR I move the flag checking, vehicle checking and creating scent over slime to private map functions.
In order not to duplicate the submap lookup code, I created a neat templated function that invokes a functor on submap tiles, intelligently going through submaps in the area rather than re-getting them each time. The function also clips area to map bounds to avoid doing bounds checks.
It could be easily re-used to speed up some other sections of code, like NPCs looking for items, creating a cache of watery tiles (to speed up field processing when there are many fields), player radiant temperature function (fires) etc.

EDIT: Tested

kevingranade commented on src/map.cpp:
```diff
@@ -6992,3 +6992,138 @@ void map::creature_on_trap( Creature &c, bool const may_avoid )
+    }
+}
+
+void map::scent_blockers( bool blocks_scent[SEEX * MAPSIZE][SEEY * MAPSIZE],
```
I think you want '''bool (&blocks_scent)[SEEX * MAPSIZE][SEEY * MAPSIZE]''', otherwise it degrades to a pointer with no size information.

Coolthulhu commented:
Corrected the pointers.
I also figured a better way to process slime: moved it to `field.cpp`, so it is like all other fields. That way it will only ever be processed when it is actually on the map.

kevingranade commented:
Definite improvement, update_scent() moved from about 8.4% of processing time to 3.13%, with 2% of that being in scent_blockers().
There's also an interesting piece of information in that breakdown, only 1/3 of the processing time is the scent diffusion code itself, so it might actually be reasonable to run diffusion over a number of different scent maps, since the work of retrieving the scent blockers can be shared.
